docs/design/script-runner.md

```diff
@@ -0,0 +1,100 @@
+* A script could depend on the result of other operations.
+* Other operations could depend on the result of a script.
+
+### API Object
+
+To achieve our design purposes, we extend the `Operation` API for users to integrate scripts into an operation pipeline.
+
+```go
+// OperationSpec defines the desired state of Operation.
+type OperationSpec struct {
+    // Processor describes how to register a operation processor into kubediag.
+    Processor Processor `json:"processor"`
+    // ......
+}
+
+// Processor describes how to register a operation processor into kubediag.
+type Processor struct {
+    // One and only one of the following processor should be specified.
+    // HTTPServer specifies the http server to do operations.
+    // +optional
+    HTTPServer *HTTPServer `json:"httpServer,omitempty"`
+    // ScriptRunner contains the information to run a script.
+    // +optional
+    ScriptRunner *ScriptRunner `json:"scriptRunner,omitempty"`
+    // ......
+}
+
+// ScriptRunner contains the information to run a script.
+type ScriptRunner struct {
+    // Script is the content of shell script.
+    Script string `json:"script"`
+    // ArgKeys contains a slice of keys in parameters or operationResults. The script arguments are generated
+    // from specified key value pairs.
+    // No argument will be passed to the script if not specified.
+    // +optional
+    ArgKeys []string `json:"argKeys,omitempty"`
+    // OperationResultKey is the prefix of keys to store script stdout, stderr or error message in operationResults.
+    // Execution results will not be updated if not specified.
+    // +optional
+    OperationResultKey *string `json:"operationResultKey,omitempty"`
+    // PodReferenceKey indicates the pod if the script need to be executed inside a pod.
+    // The value should fulfill the scheme of "namespace/name/container" or "namespace/name".
+    // The script will be run on node if not specified.
+    // +optional
+    PodReferenceKey *string `json:"podReferenceKey,omitempty"`
+}
+```
+
+The `ScriptRunner` processor is designed for users to define how should KubeDiag run their scripts. It contains the following fields:
+
+* `.spec.processor.scriptRunner.script`: The content of shell script.
+* `.spec.processor.scriptRunner.argKeys`: ArgKeys contains a slice of keys in parameters or operationResults. The script arguments are generated from specified key value pairs. No argument will be passed to the script if not specified.
+* `.spec.processor.scriptRunner.operationResultKey`: OperationResultKey is the prefix of keys to store script stdout, stderr or error message in operationResults. Execution results will not be updated if not specified.
+* `.spec.processor.scriptRunner.podReferenceKey`: PodReferenceKey indicates the pod if the script need to be executed inside a pod. The value should fulfill the scheme of "namespace/name/container" or "namespace/name". The script will be run on node if not specified.
```

Comment:
`podReferenceKey` seems like the kind of parameters that should be specified in `diagnosis`.

Reply:
This is reasonable to specify `podReferenceKey` in Diagnosis if we restrict the scope of a Diagnosis on a node. However, I am considering to extend the scope to multiple nodes since it is a common case for an operator to collect information on other nodes when resolving network problems. I will remove the `podReferenceKey` field in Operation for now and backlog such scope issue.